front/gatsby/src/components/Layout/Validators/CurrentElected.tsx

```diff
@@ -0,0 +1,165 @@
+
+import { CURRENT_ELECTED, OFFLINE_VALIDATORS } from '../graphql';
+import { OfflineValidator, Validator } from './types';
+
+interface Props {
+  sessionIndex: number;
+}
+
+interface JoinValidatorOffline extends Validator {
+  wasOfflineThisSession: boolean;
+}
+
+export const CurrentElectedList = (props: Props): React.ReactElement => {
+  const { sessionIndex } = props;
+  const { api } = useContext(ApiContext);
+
+  const [currentElected, setCurrentElected] = useState<
+    JoinValidatorOffline[]
+  >();
+
+  const currentValidators = useQuery(CURRENT_ELECTED, {
+    variables: { sessionIndex },
+    pollInterval: 5000,
+  });
+
+  const currentOffline = useQuery(OFFLINE_VALIDATORS, {
+    variables: { sessionIndex },
+    pollInterval: 5000,
+  });
+
+  useEffect(() => {
+    if (currentValidators.data && currentValidators.data.validators) {
+      const result: JoinValidatorOffline[] = [];
+
+      /*
+       * in theory not great, but unnoticeable in practice
+       * O(N*M) where N = |validators|, M = |offline|
+       */
+      if (currentOffline.data && currentOffline.data.offlineValidators) {
+        currentValidators.data.validators.map((validator: Validator) => {
+          currentOffline.data.offlineValidators.map(
+            (offline: OfflineValidator) => {
+              if (
+                validator.stash === offline.validatorId ||
+                validator.controller === offline.validatorId
+              ) {
+                result.push({
+                  ...validator,
+                  wasOfflineThisSession: true,
+                });
+              } else {
+                result.push({
+                  ...validator,
+                  wasOfflineThisSession: false,
+                });
+              }
+            }
+          );
+        });
+      } else {
+        currentValidators.data.validators.map((validator: Validator) => {
+          result.push({
+            ...validator,
+            wasOfflineThisSession: false,
+          });
+        });
+      }
+      setCurrentElected(result);
+    }
+  }, [currentValidators, currentOffline]);
+
+  const handleAddToCart = (): void => {
+    // do nothing for now
+  };
+
+  const renderValidatorsTable = (): React.ReactElement => {
+    return (
+      <Table celled collapsing padded='very' striped size='large' width='100%'>
+        <Table.Header fullWidth>
+          <Table.Row>
+            <Table.HeaderCell> Offline </Table.HeaderCell>
+            <Table.HeaderCell>Stash</Table.HeaderCell>
+            <Table.HeaderCell>Controller</Table.HeaderCell>
+            <Table.HeaderCell>Commission</Table.HeaderCell>
+            <Table.HeaderCell> </Table.HeaderCell>
+          </Table.Row>
+        </Table.Header>
+        <Table.Body>
+          {currentElected ? (
+            currentElected.map(
+              ({ stash, controller, preferences, wasOfflineThisSession }) => (
+                <Table.Row textAlign='center' key={shortid.generate()}>
```

Comment:
key is used for caching and optimization while re-rendering, adding this effectively removes all that.

I think `key={stash}` could be good? stash should be unique by row.

Reply:
up till recently that's what i would have done, however reading this suggests to use `shortid` https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318

Reply:
upon further reading though, you're right, this is also an antipattern...sigh. will use stashid instead.

Reply:
shite, so there are duplicate stashes. that should not be possible...

Reply:
ok so i'll keep shortid for now, #190 should make sure to definitively remove any duplicates.